### DISCUSSION TITLE
Can auto layout handle non-tree structures?

### DISCUSSION THREAD

**Discussion by Vazerthon** [xyflow/xyflow#2788] [Layouting] [answered]
Hi,
I'm currently evaluating Reaflow and React Flow for a project. Some of my requirements are automatic layout, structures that may have multiple root nodes and nodes with multiple parents (graphs, essentially).

I've followed [the guide for auto layout ](https://pro.reactflow.dev/examples/auto-layout) and it works great for some examples but it makes a lot of assumptions about the structure of the flow being a tree.

Is it possible to use a graph structure? If so, how?

Fwiw this is something that Reaflow handles pretty nicely (see gif). React Flow errors with `Uncaught Error: multiple roots`

![Peek 2023-01-31 11-24](https://user-images.githubusercontent.com/8657986/215764580-7922f21b-ca32-484e-bcd4-68a4f0a3803b.gif)

Thanks!

**Comment by moklick** [comment, marked as answer]
Reaflow uses [Elkjs](https://github.com/kieler/elkjs) under the hood. You can use Elkjs with React Flow too. We decided against adding a layouting library so that you can decide if you want to use [Elkjs](https://github.com/kieler/elkjs), [dagre](https://github.com/dagrejs/dagre), [d3-hierarchy](https://github.com/d3/d3-hierarchy) or something else.

**Comment by Vazerthon** [comment]
Thanks @moklick, I'll take a look at those options. Are there any official guides for working with elk and React Flow together? Dagre appears to be unmaintained so it's not really a contender

**Reply by moklick** [reply]
dagre is currently unmaintained but it's still a very good option. We have examples for d3-hierarchy and dagre but it's the same for elk, d3-hierarchy, graphology or whatever lib you are using. You wait until the nodes are initialized, put them with the edges in the layouting algorithm and pass them back to React Flow.

**Reply by Vazerthon** [reply]
Cool, I'll take a look. Thanks for your help!
